File: hash_table.py
class HashTable(object):
    def __init__(self, length = 4):
        self.array = [None] * length
    
    def hash(self, key):
        length = len(self.array)
        print("Hash Key: %s" % hash(key))
        return hash(key) % length

    def add(self, key, value):
        if self.is_full():
            self.double()

        index = self.hash(key)
        if self.array[index] is not None:
            for kvp in self.array[index]:
                if kvp[0] == key:
                    kvp[1] = value
                    break
                else:
                    self.array[index].append([key, value])
        else:
            self.array[index] = []
            self.array[index].append([key, value])

    def get(self, key):
        index = self.hash(key)
        if self.array[index] is None:
            raise KeyError()
        else:
            for kvp in self.array[index]:
                if kvp[0] == key:
                    return kvp[1]
            raise KeyError()

    def is_full(self):
        items = 0
        for item in self.array:
            if item is not None:
                items += 1
        return items > len(self.array) / 2

    def double(self):
        ht2 = HashTable(length = len(self.array) * 2)
        for i in range(len(self.array)):
            if self.array[i] is None:
                continue

            for kvp in self.array[i]:
                ht2.add(kvp[0], kvp[1])

        self.array = ht2.array

ht = HashTable()
ht.add('Luca', 1)

File: hashtable.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#define MAX 10

// key value pair struct
struct item{
    int key;
    int value;
};

//flag 0 = never used, flag 1 = currently used, flag 2 = used before
struct hashtable_item{
    int flag;
    struct item *data;
};

// array struct
int size = 0;
struct hashtable_item array[MAX];

// init array with 0 flags
void init_array(){
    for (int i = 0; i < MAX; i++){
        array[i].flag = 0;
        array[i].data = NULL;
    }
}

// hash f
int hash(int key){
    return (key % MAX);
}

// add module
void add(key, value){
    // hashed index
    int index = hash(key);
    int i = index;

    // memory alloc for new item + assignments
    struct item *new_item = (struct item*)malloc(sizeof(struct item));
    new_item->key = key;
    new_item->value = value;

    // if [i] is occupied
    while (array[i].flag == 1){
        // if key already exists
        if (array[i].data->key == key){
            printf("Updating %d\n", key);
            // update its value
            array[i].data->value = value;
            return;
        }
        // else linear-probe through indexes
        i = (i + 1) % MAX;
        // if every index is occupied
        if (i == index){
            printf("Hash table is full\n");
            return;
        }
    }

    // assignments
    array[i].flag = 1;
    array[i].data = new_item;
    size++;
    printf("Key %d has been inserted\n", key);
    return;
}

// exists O(1)
bool exists(key){
    if (array[hash(key)].flag == 1)
        return true;
    else return 0;
}

// get value by key
int get(key){
    if (exists(key))
        return array[hash(key)].data->value;
    else return 0;
}

void ht_remove(key){
    if (exists(key)){
        array[hash(key)].data = NULL;
        array[hash(key)].flag = 2;
        printf("Successfully removed (%d) key\n", key);
        return;
    }
    else
        printf("No such key in HT\n");
}

void printTable(){
    printf("HashTable(k,v): {\n");
    for (int i = 0; i < MAX; i++)
        if (array[i].data != NULL)
            printf("(%d, %d)\n", array[i].data->key, array[i].data->value);
    printf("}\n");
}

void main(){
    init_array();
    int choice, key, value;

    printf("Hash Table initialized\n");

    for(;;){
        printf("\n\nChoose a method\n1. Add key-value\n2. Remove by key\n3. Check existence\n4. Get value by key\n5. Print Table\n6. Quit\n\n\n");
        scanf("%d", &choice);
        switch(choice){
            case 1:
                printf("Choose key and value\n");
                scanf("%d %d", &key, &value);
                add(key, value);
                break;
            case 2:
                printf("Choose key to remove\n");
                scanf("%d", &key);
                ht_remove(key);
                break;
            case 3:
                printf("Choose key to check\n");
                scanf("%d", &key);
                exists(key);
                break;
            case 4:
                printf("Choose key to get value from\n");
                scanf("%d", &key);
                if(get(key))
                    printf("(%d) value is (%d)\n", key, get(key));
                else
                    printf("No such key in HT\h");
                break;
            case 5:
                printTable();
                break;
            case 6:
                return 0;
            default:
                printf("Insert a valid entry\n");
                break;
        }
    }
    return 0;
}
